tutorials/14-minituring/14_minituring.jmd

```diff
@@ -0,0 +1,327 @@
+
+```julia; eval=false
+@mini_model function m(x)
+    a ~ Normal(0.5, 1)
+    b ~ Normal(a, 2)
+    x ~ Normal(b, 0.5)
+end
+```
+
+Specifically,
+
+  - All observed variables should be arguments of the program.
+  - No control flow is permitted in the model definition.
+  - All variables should be scalars.
+  - No function return.
+
+```julia
+# First, we import some needed packages
+using MacroTools, Distributions, Random, AbstractMCMC, MCMCChains
+```
+
+Before getting to the actually compiler, let's first build the data structure for program trace.
+A program trace for a probabilistic programming language need to at least record the values of stochastic variables and (log)probability of the corresponding stochastic variables scored by the prior distributions.
+
+> **Julia Sidebar**
+> Julia's high-performance is relied upon its multiple-dispatched based compilation scheme.
+> In so many words, Julia compiler will produce a compiled version of the program customized for some specific types.
+> Thus, in the idea case, it should have very similar performance as a pre-compiled C program.
+> The implication of this is that although typing is not necessary for correct program execution, performance will heavily rely on programmer providing correct and coherent type information.
+
+```julia
+struct VarInfo{V,L}
+    values::V
+    logps::L
+end
+
+VarInfo() = VarInfo(Dict{Symbol,Float64}(), Dict{Symbol,Float64}())
+
+function Base.setindex!(varinfo::VarInfo, (value, logp), var_id)
+    varinfo.values[var_id] = value
+    varinfo.logps[var_id] = logp
+    return varinfo
+end
+```
+
+Now, let's define a `context` that does something very simple: if a stochastic variable is observed, we compute its log probability; or, if it is assumed(sampled), then we sample from prior and also computes its log probability.
+`context` is an Turing internal implementation mechanism.
+Because different inference algorithms may have slightly different actions to do with stochastic variables, we want a good abstraction and interface to implement these.
+The name `context` comes from the "contextual dispatching" idea of [Cassette.jl](https://github.com/JuliaLabs/Cassette.jl).
+Every `context`s is defined to be a type in Julia, and the `observe` and `assume` functions will take a `AbstractContext` type as their first arguments.
+Then, at runtime, the Julia multiple-dispatch system will dispatch corresponding functions according to the type that is a specific `context`.
+
+**Note:** *Although the context system in this tutorial is inspired by Turing's, Turing's context system is battle-tested and much more complicated.
+Thus, readers are advised to refer to Turing's documentations and code if knowledge of Turing's context system is required.
+Also, it should be pointed out that the context system design in Turing is not in its definitive form as of the current version at the time of this notebook is written.*
+
+```julia
+abstract type AbstractContext end
```

Comment:
I think keeping the `MHTransitionContext` has its benefits as it demonstrates the idea of using a `context` to modify the behavior of `assume` and `observe`.

Reply:
`SamplingContext` is already an example for that, isn't it? And you can still specialize `assume` and `observe` based on the sampler (which is how it is done in Turing anyway).